### DISCUSSION TITLE
why the JsonServiceClient not use HttpClient?

### DISCUSSION THREAD

**Discussion by 1257960069** [ServiceStack/Discuss#103] [Q&A] [answered]
from source code of AsyncServiceClient.I found it 
#if NET6_0_OR_GREATER
        public System.Net.Http.HttpClient HttpClient { get; set; }
#endif

but it just use  the extensions method (var webReq = this.CreateHttpWebRequest(requestUri)) at method SendWebRequestAsync
it seems that HttpClient  is not usefull, the usefull system http class is HttpWebRequest.

how to control JsonServiceClient's Httpclient lifetime?

**Comment by mythz** [comment]
The newest C# ServiceClient that uses HttpClient you should use is the `JsonApiClient` which for .NET 6+ is included in the **ServiceStack.Client** NuGet package, see:

https://docs.servicestack.net/csharp-client#jsonapiclient

```csharp
var client = new JsonApiClient(baseUri);
```

Which can also be registered in your App's IOC using the HttpClient factory:

```csharp
builder.Services.AddJsonApiClient(builder.Configuration["BaseUrl"]);
```

Which supports both Async and Sync APIs.

The older `JsonServiceClient` is for legacy clients which uses .NET Framework's older HttpWebRequest implementation.

**Reply by 1257960069** [reply]
sorry,I just view at Service Gateway. not view new docs.
there are some issue about ServiceStack.Discovery.Consul at version 8.0

System.TypeLoadException:“Could not load type 'ServiceStack.MetadataFeatureExtensions' from assembly 'ServiceStack, Version=6.0.0.0, Culture=neutral, PublicKeyToken=null'.”

https://docs.servicestack.net/service-discovery#servicestack.discovery.consul
<PackageReference Include="ServiceStack.Discovery.Consul" Version="6.*" />
![image](https://github.com/ServiceStack/Discuss/assets/31265244/8efb4db0-3517-40c2-9301-f3d39a673433)

it seems not update yet for long time, so need remove it from offical docs ?

**Reply by mythz** [reply]
I've removed a link to the main menu and updated the [Service Discovery](https://docs.servicestack.net/service-discovery) page to include a warning that the projects are deprecated however they should still be useful in showcasing how to implement a similar service discovery feature.

**Reply by 1257960069** [reply]
public class MyServices2 : Service
 {
     public async Task<object> Any(GetCustomer request)
     {
         //https://docs.servicestack.net/service-gateway#service-integration-apis

         //MyServices instance from Container,it should dispose by Container's litetime
         //should we dispose it by using?
         using var myService = base.ResolveService<MyServices>();
         var r = (HttpResult)myService.Any(new Hello { Name = request.Id.ToString() });

         //not found ExecuteRequest method from base Service 
         //  var  Customer = base.ExecuteRequest(new Hello { Name = request.Id .ToString()}),

         //the dto is Hello : IReturn<HelloResponse>
         //can not get result from object if return type is HttpResult, even new HttpResult(new HelloResponse) 
         //should we need enhance the code to check the HttpResult.Response
         var r1 = await Gateway.SendAsync(new Hello { Name = request.Id.ToString() });
         //the way can work 
         var r2 = await Gateway.SendAsync<HttpResult>(new Hello { Name = request.Id.ToString() });



         return r;
     }
 }
![image](https://github.com/ServiceStack/Discuss/assets/31265244/c6c4f6aa-029e-404a-97b5-0db2919ac3c8)

**Reply by mythz** [reply]
The Gateways only return typed Response DTO's not `HttpResult`, which is used to return a decorated Response to customize its HTTP API Responses, which aren't used in Gateway requests which only return an APIs Typed Response DTOs. Typically Gateway Requests are in-process which avoid HTTP and Serialization. 

Make sure your Request DTOs specify their response type on the Request DTO with `IReturn<HelloResponse>` that way you don't have to specify it on the call site, e.g:

```csharp
HelloResponse response = await Gateway.SendAsync(new Hello { ... });
```

**Reply by 1257960069** [reply]
MyServices instance resoved from Container,it should be dispose by Container's litetime automatically.
using var myService = base.ResolveService<MyServices>();
should we dispose it by using?

**Reply by mythz** [reply]
Yes, if you manually resolve a service you should also dispose of it as you're doing with `using var`

**Reply by 1257960069** [reply]
I don't think so.   at asp.net core will automatically disposing.
![image](https://github.com/ServiceStack/Discuss/assets/31265244/05ed5446-b10b-4d5a-b8f8-1af87a4c14ce)
![image](https://github.com/ServiceStack/Discuss/assets/31265244/211a3acf-3601-4750-8730-bd4bd0a5eae0)

**Reply by mythz** [reply]
`ResolveService` should only resolving ServiceStack Services which need to be explicitly disposed. You would use `TryResolve<T>` to resolve dependencies from the IOC.

**Reply by 1257960069** [reply]
you are wrong. TryResolve<T> . I think this is not a good design. It's confusing
![image](https://github.com/ServiceStack/Discuss/assets/31265244/98bbb448-2beb-4ddf-b331-f5ca499621e7)
![image](https://github.com/ServiceStack/Discuss/assets/31265244/a19211d4-db3d-4143-951e-01e9a23a4d46)

**Reply by mythz** [reply]
If you're going to be rude don't bother posting here again.

I said `ResolveService<T>` is for resolving ServiceStack Services which it is, I've just added an `IService` type constraint to enforce this in future versions. `TryResolve<T>` is for resolving dependencies from the IOC. In future versions you can use the new `GetService<T>` and `GetRequiredService<T>` APIs for resolving dependencies from the IOC.

**Reply by 1257960069** [reply]
sorry, I just can speak some sample english. I just say you tell me things is wrong. not you are wrong. haha...

**Reply by 1257960069** [reply]
but it is not sove the "dispose" issue

**Reply by mythz** [reply]
You don't need to dispose dependencies registered in ASP.NET Core IOC, although it's recommended to use constructor injection for any dependencies you need to avoid manually resolving them, e.g:

```csharp
public class MyServices(WheatherForecast forecast) : Service
{
    public object Any(Hello request)
    {
        return forecast.MyMethod()
    }
}
```

**Reply by 1257960069** [reply]
I know that, But sometimes resove service in methods can avoid over dependence. like [FromService] 
 public class WeatherForecastController : ControllerBase
 {
     private readonly ILogger<WeatherForecastController> _logger;

     public WeatherForecastController(ILogger<WeatherForecastController> logger)
     {
         _logger = logger;
     }

     [HttpGet(Name = "GetWeatherForecast")]
     public string Get([FromServices] WeatherForecast weatherForecast )
     {
         return "ok";
     }

     [HttpGet(Name = "GetWeatherForecast2")]
     public string Get2()
     {
         return "ok";
     }
 }

**Reply by 1257960069** [reply]
use constructor injection,but not dispose yet at request end.
I think this could lead to memory leaks, and if each service needs to be manually disposing, it would lose the convenience of automatic disposing through dependency injection
![image](https://github.com/ServiceStack/Discuss/assets/31265244/9b72f003-0d0f-45d1-adcf-2c5dc11f3820)
![image](https://github.com/ServiceStack/Discuss/assets/31265244/6686c2ab-0aad-482e-8709-49fcb581fd94)

**Reply by mythz** [reply]
Since you prefer to use ASP.NET Core IOC you can make use of our new Endpoint Routing and IOC feature which makes use of ASP.NET Core's IOC for registering and resolving Services and ASP.NET Core Endpoint Routing for calling them.

What template is your project based on? I'll convert it to use the new Endpoint Routing support now available in our pre-release packages

**Reply by 1257960069** [reply]
I use command "x new web MyWeb" to create the template

**Reply by mythz** [reply, marked as answer]
ok I've got an updated project which you can create with:

```bash
x new mythz/web MyWeb
```

It uses ASP.NET Core's IOC for all dependencies (inc. Services), uses Endpoint Routing and System.Text.Json by default.
I've also configured it with ASP.NET Core's Swagger UI so you can see that ServiceStack Services now integrates with other ASP.NET Core Minimal APIs, controllers etc which also use endpoint routing.

We're currently working on the docs now, but should be hopefully ready for release by next week.

**Reply by 1257960069** [reply]
Thank you for your immediate feedback.  asp.net core servicecollection can not support property in the past, but now can work at property,however,it need a attribute named [FromService].  
![9f0bfbcddfac1ec733a5ce5501a0f2c](https://github.com/ServiceStack/Discuss/assets/31265244/bfa6e87b-6358-4d38-97cd-ef49f830aee5)
